FastAPI conversion

Kilo59 opened the pull request:
I thought I should open up my in-progress PR so that the maintainers can see what a transition to 
#130  FastAPI would look like.

I didn't want to heavily redesign the application without input or discussion from the original creators.
But for example, some of the existing code could be refactored to take advantage of FastAPI's [dependency injection system](https://fastapi.tiangolo.com/tutorial/dependencies/).

I've added 2 `pipenv` scripts, that can be used to run a local development server.
To run it just type. 
`pipenv run dev_app` or `pipenv run app`

### TODO:
- [x] fix timelines for `GET /locations`
- [x] refactor according to guidance from creators
- [x] work with `csbs` source 
- [x] verify working `/v1` endpoints
- [x] setup gunicorn configuration
- [x] Test Heroku deployment
- [x] add GET /sources
- [x] fix datetime format
- [x] add `source` param to all endpoints?
- [x] make `timelines` query param a boolean (in swagger doc)
- [x] replace flask v2 with FastAPI v2
- [x] FastAPI app tests
- [ ] Make all properties on a location queryable

Kilo59 commented:
Swagger/OpenAPI docs

![image](https://user-images.githubusercontent.com/13108583/77260444-768e8a00-6c5e-11ea-96d5-c6705292226a.png)

ReDocs
![image](https://user-images.githubusercontent.com/13108583/77260505-c8371480-6c5e-11ea-815f-4de387654564.png)


![image](https://user-images.githubusercontent.com/13108583/77260483-ae95cd00-6c5e-11ea-8190-af3bc7f88f95.png)

ExpDev07 commented:
Looks good. I don't like that everything is in ``main.py``. Hopefully we can for example put models under ``/models``, etc.

Kilo59 commented:
> Looks good. I don't like that everything is in `main.py`. Hopefully we can for example put models under `/models`, etc.

Yes absolutely, I planned on refactoring and moving those into a `models.py`, just thought I'd talk to you first about exactly how you'd like that done.

ExpDev07 commented:
> just thought I'd talk to you first about exactly how you'd like that done.

That's considerate of you, thanks a lot. I really like dividing it up as much as possible, so you have the ``/models`` directory, then ``/models/location.py``, ``/models/timeline.py``, etc.

Kilo59 commented on app/main.py:
```diff
@@ -0,0 +1,124 @@
+
+
+# mount the existing Flask app to /v2
+APP.mount("/v2", WSGIMiddleware(create_app()))
```
I'm not sure how this behaves with the `/v1` endpoints. At least locally, the `/v2` endpoints appear to be functional, including the redirect to your github page when hitting `/v2`.

ExpDev07 replied:
So FastAPI works with flask or? The v1 endpoints can be found at:
1. https://coronavirus-tracker-api.herokuapp.com/all
2. https://coronavirus-tracker-api.herokuapp.com/confirmed
3. https://coronavirus-tracker-api.herokuapp.com/deaths
4. https://coronavirus-tracker-api.herokuapp.com/recovered

I have checked the logs, and they are still in use by a lot of apps, so they need to work.

Kilo59 replied:
@ExpDev07 
So because /v1 and /v2 are defined into the same app, this appears to put both `/v2` and `/v1` behind the `/v2` prefix.

In order to separate them I believe I need to create 2 separate flask apps and mount them separately.

For example `/confirmed` is part of `/v1`, but I can access it locally at `/v2/confirmed` when it should be `/v1/confirmed`.

Kilo59 replied:
I'll probably need to do something like...
```python
APP.mount("/v1", WSGIMiddleware(create_app("v1")))
APP.mount("/v2", WSGIMiddleware(create_app("v2")))
```

ExpDev07 replied:
It should be ``/confirmed`` without any prefix :). Hmm, two different flask apps? FastAPI doesn't support blueprinting?

ExpDev07 replied:
@Kilo59 it looks like it supports different routers. Look here: https://www.errietta.me/blog/python-fastapi-intro/, and search for ``include_router``.

Kilo59 replied:
> So FastAPI works with flask or?

@ExpDev07 
Yes, it does.
https://fastapi.tiangolo.com/advanced/wsgi/
My thinking is that we could start deploying with FastAPI with versions 1 and 2 ["mounted"](https://fastapi.tiangolo.com/advanced/wsgi/).
 
And with a FastAPI version of `/v2` that lives at `/v2-1` (or `/v2-beta`), once we have confirmed that it works under stress and smoothed out any edge-cases, we could remove the flask mounted v2 (and keep mounting v1) and move the `FastAPI` version to `/v2`.

Hopefully, that made sense 😄

Kilo59 replied:
> @Kilo59 it looks like it supports different routers. Look here: https://www.errietta.me/blog/python-fastapi-intro/, and search for `include_router`.
> 
> So we'll just have a different `APIRouter` per api version. Something like:
> Seems to be exactly what we're looking for.

Yes, we can do that too, but then we'd have to rewrite those versions with FastAPI instead of just mounting them and forgetting about it.
Which we definitely could do, as it would mean they could have higher throughput and their own swagger documentation.

ExpDev07 replied:
I mean if the old v1 enpoints work then, then I guess it's fine.

Kilo59 replied:
I'd be happy to get on a screen share, discord etc. to talk about this stuff (and deployment configuration, etc.).

ExpDev07 replied:
My Discord # is: Marius Big Ounce#2997.

Kilo59 replied:
Just sent you an invite.
Kilo59#8819

ExpDev07 replied:
I think I accidentally declined it, sent you one now.

Kilo59 commented:
Here's a response comparison between the currently deployed application and this one.
The only difference is the format of `last_updated`.

GET `/locations?country_code=IT`

---------------------

### v2
```json
{
  "latest": {
    "confirmed": 53578,
    "deaths": 4825,
    "recovered": 6072
  },
  "locations": Array[1][
    {
      "coordinates": {
        "latitude": "43",
        "longitude": "12"
      },
      "country": "Italy",
      "country_code": "IT",
      "id": 16,
      "last_updated": "2020-03-22T22:02:46.078243Z",
      "latest": {
        "confirmed": 53578,
        "deaths": 4825,
        "recovered": 6072
      },
      "province": ""
    }
  ]
}
```

-------------

### v2-1
```json
{
  "latest": {
    "confirmed": 53578,
    "deaths": 4825,
    "recovered": 6072
  },
  "locations": Array[1][
    {
      "coordinates": {
        "latitude": "43",
        "longitude": "12"
      },
      "country": "Italy",
      "country_code": "IT",
      "id": 16,
      "last_updated": "2020-03-22T22:12:42.421409+00:00",
      "latest": {
        "confirmed": 53578,
        "deaths": 4825,
        "recovered": 6072
      },
      "province": ""
    }
  ]
}
```

ExpDev07 commented:
Awesome, how about the v1 endpoints?

Kilo59 commented:
Test deployment on heroku

OpenAPI
https://covid-tracker-beta.herokuapp.com/v2-beta

Docs 
https://covid-tracker-beta.herokuapp.com/docs

ExpDev07 commented on Pipfile:
```diff
@@ -9,13 +9,19 @@ pytest = "*"
 flask = "*"
 python-dotenv = "*"
 requests = "*"
 gunicorn = "*"
```
Can be removed now that we're using uvicorn.

Kilo59 replied:
Actually we are running `uvicorn` workers via `gunicorn`.
https://www.uvicorn.org/deployment/

Kilo59 replied:
https://fastapi.tiangolo.com/deployment/#alternatively-deploy-fastapi-without-docker

Kilo59 commented:
> There's an issue where when using alternative data-sources for locations, their fields won't show up. For example, `/v2/locations?source=csbs` won't display the `county` field as defined in `app/location/csbs.py - CSBSLocation`.

Fixed it by adding the `county` field to the mode.
Also added some very minimal tests of the `v2/location` endpoint itself.
I certainly can add more.

ExpDev07 commented:
To start the app, now use

* ``pipenv run dev`` for development (with live-reload), and
* ``pipenv run start`` for starting without livereload.

ExpDev07 commented:
@all-contributors please add @Kilo59 for code, infrastructure, testing, and documentation.
